_posts/2022-04-21-nextjs_frontend_framework_with_be_api.md

```diff
@@ -0,0 +1,151 @@
+---
+layout: post
+title: How to setup NextJS FrontEnd Framework with some API? 
+categories: [NextJS]
+author_github: nevermore17
+author: Verbin Kirill
+---
+
+Let's say we have some BackEnd server (in my case, Django + Django Rest Framework) and want to generate `html`/`css` on NextJS
+
+## 1. Create NextJS app
+ ``` 
+    npx create-next-app myapp
+ ```
+NextJS application using a folder structure such as:
+```
+myapp
+|-- pages
+|   |-- api
+|   |   |-- hello.js
+|   |
+|   |-- _app.js
+|   |-- index.js
+|
+|-- public
+|   |-- favicon.ico
+|  
+|-- styles
+|   |-- global.css
+|   |-- Home.module.css
+|
+|--package.json
+
+```
+On this step I recommend add to myapp folder files `.env.development` and `.env.production`. Depending on the start mode(dev or prod) NextJs add values from `.end.<start_mode>` files to `process.env`. This will make development and deployment easier.
+``` env
+# .env.development
+
+NEXT_PUBLIC_BACKENDSERVER=http://127.0.0.1
+NEXT_PUBLIC_BACKENDPORT=8000
+```
+## 2. Create page articles
+
+Paths for the application are made using folder management in pages folder:
+- pages/index.js - `/`
+- page/some_page - `/some_page`
+- page/some_page/some_page2.js - `/some_page/some_page2`
+- page/[param].js - `/<param>`
+- page/[param]/[param2] - `/<param>/<param2>`
+
+We have to create `/articles` way. 
+```js
+// pages/articles.js
+
+export default function Index() {
+    return(
+        <>
+            Hello World!
+        </>
+    )
+}
+```
+And run app. Now we have "Hellow world" on rout `/articles`. But we need get articles from BackEnd server
+
+In myapp BackEnd server have 2 routes:
+- `http://127.0.0.1:8000/api/article/`
+- `http://127.0.0.1:8000/api/article/<int:pk>`
+
+Page `/article/` only lists links to article pages. 
+Function `getServerSideProps` without params request articles from BackEnd server and return like page_props
```

Comment:
The function `getServerSideProps` without parameters requests articles from BackEnd server and returns as page_props